### PR TITLE
Move multiple-port ServerImpl to NettyServer

### PR DESCRIPTION
**Pull request by YifeiZhuang**
This is the first step towards #6641 , a deadlock issue at ServerImpl.start().
It makes ServerImpl to have a single transport server, and this single transport server (NettyServer) will bind to all listening addresses during bootstrap, as supported by Netty 4.x.

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/ServerImpl.java [hunk 1]
```diff
@@ -388,8 +378,7 @@ public void serverShutdown() {
       ArrayList<ServerTransport> copiedTransports;
       Status shutdownNowStatusCopy;
       synchronized (lock) {
-        activeTransportServers--;
-        if (activeTransportServers != 0) {
+        if (transportServersTerminated) {
```
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/ServerImpl.java:381]: Maybe `if(serverShutdownCallbackInvoked)`?

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 2]
```diff
@@ -144,32 +157,49 @@
...
-    return channel.localAddress();
+  }
+
+  @Override
+  public List<SocketAddress> getListenSocketAddresses() {
+    List<SocketAddress> listenSocketAddresses = new ArrayList<>();
+    for (Channel c: channelGroup) {
+      listenSocketAddresses.add(c.localAddress());
+    }
+    return listenSocketAddresses;
```
Comment (dapengzhang0) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:182]: Why is this not consistent with `getListenSocketAddress()` for the case "server is not listening/bound yet"

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 3]
```diff
@@ -144,32 +157,49 @@
...
   @Override
   public void start(ServerListener serverListener) throws IOException {
     listener = checkNotNull(serverListener, "serverListener");
 
-    ServerBootstrap b = new ServerBootstrap();
+    final ServerBootstrap b = new ServerBootstrap();
     b.option(ALLOCATOR, Utils.getByteBufAllocator(forceHeapBuffer));
     b.childOption(ALLOCATOR, Utils.getByteBufAllocator(forceHeapBuffer));
-    b.group(bossGroup, workerGroup);
+    b.group(bossExecutor, workerGroup);
```
Comment (dapengzhang0) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:202]: What's the reason for this change?

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 4]
```diff
@@ -258,51 +288,88 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
+            return map;
+          }
+        }
+    );
+    Map<ChannelFuture, SocketAddress> channelFutures;
+    try {
+      Future<Map<ChannelFuture, SocketAddress>> bindCallResult =
+          bindCallFuture.awaitUninterruptibly();
+      if (!bindCallResult.isSuccess()) {
+        throw new IllegalStateException(bindCallResult.cause());
```
Comment (dapengzhang0) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:310]: Why not just throw `bindCallResult.cause()`?

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 5]
```diff
@@ -258,51 +288,88 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
+      }
+      channelFutures = bindCallResult.get();
+    } catch (Exception ex) {
+      throw new IOException(String.format("Failed to bind to addresses %s",
+          addresses), ex.getCause());
```
Comment (dapengzhang0) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:315]: Why not just `ex` rather than `ex.getCause()`?

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 6]
```diff
@@ -226,7 +256,7 @@ public void initChannel(Channel ch) {
         ServerTransportListener transportListener;
         // This is to order callbacks on the listener, not to guard access to channel.
         synchronized (NettyServer.this) {
-          if (channel != null && !channel.isOpen()) {
+          if (terminated) {
             // Server already shutdown.
```
Comment (dapengzhang0) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:260]: terminated is slightly different from shutdown. `shutdown` means no more new stream is allowed but the existing streams can still be active until they finish. `terminated` is once `shutdown` _and_ all streams are closed.  Here the comment `// Server already shutdown` seems to have been not accurate, whereas the old comment [here L283](https://github.com/grpc/grpc-java/pull/7674/files#diff-4cf42ed83c97f1bb5a16f70fce7d0a6719bb57c68e31a6c27aa2336fc5da53aeL283) `// Already closed` seems to be correct.

File: netty/src/main/java/io/grpc/netty/NettyServer.java (hunk 3 above)
Comment (YifeiZhuang) [reply netty/src/main/java/io/grpc/netty/NettyServer.java:202]: @ejona86  suggested before(https://github.com/grpc/grpc-java/issues/6641#issuecomment-734397031), and I thought it is a brilliant idea in terms of thread safety:
> If NettyServer is handling the multiple ports, then we can do all the binds as one "atomic" operation on the event loop. This depends on a behavior of Channel that it executes the operation immediately if already on the event loop.

**Comment by dapengzhang0** [comment]
@YifeiZhuang The way you merge your PR with master head is not correct because the diff does not only show your change. The right way should be `git pull https://github.com/grpc/grpc-java.git master`.  I suggest do the following to correct it:

1. `git revert [last commits of other authors]`
2. `git pull https://github.com/grpc/grpc-java.git master`
3. [If there are any merge conflicts, resolve them manually, then `git add -u` then `git commit`]
4. `git push <your_repo> <your_branch>`

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 7]
```diff
@@ -97,10 +106,14 @@
   private final TransportTracer.Factory transportTracerFactory;
   private final InternalChannelz channelz;
   // Only modified in event loop but safe to read any time.
-  private volatile InternalInstrumented<SocketStats> listenSocketStats;
+  private volatile List<InternalInstrumented<SocketStats>> listenSocketStatsList =
+      new ArrayList<>();
+  private volatile boolean terminated;
+  private final EventLoop bossExecutor;
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:112]: Why is this a field? It looks like it should be a local variable in `start()`

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 8]
```diff
@@ -258,51 +292,88 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
+    );
+    Map<ChannelFuture, SocketAddress> channelFutures;
+    try {
+      Future<Map<ChannelFuture, SocketAddress>> bindCallResult =
+          bindCallFuture.awaitUninterruptibly();
+      if (!bindCallResult.isSuccess()) {
+        throw bindCallResult.cause();
+      }
+      channelFutures = bindCallResult.get();
+    } catch (Throwable ex) {
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:317]: Catching Throwable is generally wrong, unless it is being re-thrown unchanged. Certainly here where we wrap it in an IOException is probably a bad idea. Errors should generally propagate.

I think the `bindCallResult.cause()` should just be wrapped in an IOException immediately. The `get()` should be replaced with `getNow()` to avoid potential exceptions. At that point, there's no need for the catch.

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 9]
```diff
@@ -258,51 +292,88 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
-        listenSocketStats = new ListenSocket(channel);
-        channelz.addListenSocket(listenSocketStats);
+
+    for (Map.Entry<ChannelFuture, SocketAddress> channelFutureEntry: channelFutures.entrySet()) {
+      // We'd love to observe interruption, but if interrupted we will need to close the channel,
+      // which itself would need an await() to guarantee the port is not used when the method
+      // returns. See #6850
+      ChannelFuture future  = channelFutureEntry.getKey().awaitUninterruptibly();
+      if (!future.isSuccess()) {
+        throw new IOException(String.format("Failed to bind to address %s",
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:328]: This leaks all the other ports. When any one bind fails, we should close all the channels. Note that the channelz registration may also be impacted.

Channelz is probably easiest handled by two listeners added immediately after bind. A listener on the bindFuture would register the new socket. A listener on bindFuture.channel().closeFuture() would deregister the socket. I'll note that while closeFuture() may look really useful, you have to be careful with it, and it runs at an awkward time. I suggest only using it when the execution time doesn't matter. closeFuture() is different than "the future returned from close()".

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 10]
```diff
@@ -258,51 +292,88 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
+      channelGroup.add(channel);
+    }
   }
 
   @Override
   public void shutdown() {
-    if (channel == null || !channel.isOpen()) {
-      // Already closed.
+    if (terminated) {
+      // Already closed. Not thread safe but we are not making it worse for now.
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:347]: What does this comment mean by "making it worse"?

File: core/src/main/java/io/grpc/internal/ServerImpl.java [hunk 11]
```diff
@@ -144,7 +143,7 @@
         Preconditions.checkNotNull(builder.fallbackRegistry, "fallbackRegistry");
     Preconditions.checkNotNull(transportServers, "transportServers");
     Preconditions.checkArgument(!transportServers.isEmpty(), "no servers provided");
-    this.transportServers = new ArrayList<>(transportServers);
+    this.transportServer = transportServers.get(0);
```
[the remark this answers is not in the document]
Comment (YifeiZhuang) [reply core/src/main/java/io/grpc/internal/ServerImpl.java:146]: Because this is a two way door: it allows starting multiple servers, each supports multiple ports. And at that time, it won't break users' code again. Do you suggest to pass in only one server?

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 12]
```diff
@@ -78,7 +87,7 @@
   private EventLoopGroup bossGroup;
   private EventLoopGroup workerGroup;
   private ServerListener listener;
-  private Channel channel;
+  private final ChannelGroup channelGroup = new DefaultChannelGroup(GlobalEventExecutor.INSTANCE);
```
[the remark this answers is not in the document]
Comment (YifeiZhuang) [reply netty/src/main/java/io/grpc/netty/NettyServer.java]: This executor's job seems merely notifying the the channelGroupFuture when the operations are done. We use it to kick off the listener at close(), which won't affect the scheduling or job execution at start(). And in general yes I guess bossExecutor might be more appropriate. I got new one from .next() from the bossGroup just to decouple, and also to make test easier.

File: netty/src/main/java/io/grpc/netty/NettyServer.java (hunk 8 above)
Comment (YifeiZhuang) [reply netty/src/main/java/io/grpc/netty/NettyServer.java]: Thanks. It made it simpler and more readable now. Still we throw IOException for whatever failures that prevent us from getting the bindfuture, although it may not be from the bind itself. I guess it is somewhat unnecessary to throw different exceptions types at a higher granularity, at least for now.

File: netty/src/main/java/io/grpc/netty/NettyServer.java (hunk 9 above)
Comment (YifeiZhuang) [reply netty/src/main/java/io/grpc/netty/NettyServer.java]: It looks that channel will be closed by b.bind() if that fails and channel only be set once b.bind() future isSuccess(). So I think we can trust the result of the future instead of adding listener to clean up by ourselves. Now I collect all the successful channels and then close them all if one fails to avoid leakage. Maybe we should consider allowing partial failures in the future.

File: core/src/main/java/io/grpc/internal/ServerImpl.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (ejona86) [reply core/src/main/java/io/grpc/internal/ServerImpl.java:146]: I don't understand. Only the first server is accessed. How would code not be broken if they passed two servers?

FWIW, this is not directly called by users, but instead is called by the builder.

[the remark this answers is not in the document]
Comment (YifeiZhuang) [reply core/src/main/java/io/grpc/internal/ServerImpl.java:146]: Because it is called from builder and Netty is the only place right now we provide this list, and we guarantee it is a singleton list. So there won't be two servers passing in right now. 
What I meant was that, IIRC, we might consider supporting multiple transportServers in the ServerImpl, each of those transport servers can support multiple ports, and at that time, we would need a list of transportServers here in the constructor, instead of one object, and if we keep it as a list, we do not need refactor again at that time.

I agree the "singleton list guarantee" is less intuitive and can be confusing, as a trade off of maybe forward compatibility. I do not have strong preference over one or another, and It looks like both of you prefer passing in a single object. I can refactor it then.

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 13]
```diff
@@ -144,32 +158,57 @@
...
+    return listenSocketStatsList;
   }
 
   @Override
   public void start(ServerListener serverListener) throws IOException {
     listener = checkNotNull(serverListener, "serverListener");
 
-    ServerBootstrap b = new ServerBootstrap();
+    final ServerBootstrap b = new ServerBootstrap();
+    final EventLoop bossExecutor = bossGroup.next();;
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:208]: s/;;/;/

Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:208]: We should use the same event loop for the channels and the channelGroup. It's super-easy to and it means there are less threads involved for debugging/understanding the system.

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 14]
```diff
@@ -258,51 +297,132 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
+      throw new IOException(String.format("Failed to bind to addresses %s",
+          addresses), bindCallFuture.cause());
     }
-    channel = future.channel();
-    channel.eventLoop().execute(new Runnable() {
-      @Override
-      public void run() {
-        listenSocketStats = new ListenSocket(channel);
-        channelz.addListenSocket(listenSocketStats);
+    BindResult bindResult = new BindResult();
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:320]: Just inline the list and Throwable here, and get rid of the class? It seems the class only hides one `if` statement checking for null that could be inlined here (or simply be removed; there's nothing special about the first or last throwable).

I don't mind helpers like this. It just seems this helper isn't really helping and the indirection makes the code harder to read.

I think this loop would be simpler if you pre-added add the channels to the channelGroup; the channels exist as soon as the future is created. Then, if any future fails you close the channelGroup and throw immediately. That would remove the need for BindResult completely.

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 15]
```diff
@@ -258,51 +297,132 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
+      if (!future.isSuccess()) {
+        bindResult.failAddress(entry.getValue()).causedBy(future.cause());
+      } else {
+        final Channel channel = future.channel();
+        channel.eventLoop().execute(new Runnable() {
+          @Override
+          public void run() {
+            InternalInstrumented<SocketStats> listenSocketStats = new ListenSocket(channel);
+            channelz.addListenSocket(listenSocketStats);
+            listenSocketStatsList.add(listenSocketStats);
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:335]: This management of the `listenSocketStatsList` seems quite a bit more involved and error-prone than my suggestion of doing the equivalent of `channel.closeFuture().addListener(() -> channelz.removeListenSocket(listenSocketStats))` right here. (The code is just a sketch, we can't use lambdas.)

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 16]
```diff
@@ -258,51 +297,132 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
+          }
+        });
+        channelGroup.add(channel);
       }
-    });
+    }
+
+    if (!bindResult.isSuccess()) {
+      if (!channelGroup.isEmpty()) {
+        closeChannelGroup().awaitUninterruptibly();
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:344]: I see that closeChannelGroup() was split out to be reused here, but it doesn't seem safe to call. It does inappropriate things in this context like sharedResourceReferenceCounter.release() and protocolNegotiator.close(). (There's pre-existing issues already surrounding sharedResourceReferenceCounter, but this change doesn't fix them so it seems better to leave it alone.)

If listenSocketStatsList is removed by my other suggestion, I think this just becomes `channelGroup.close().awaitUninterruptibly()`.

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 17]
```diff
@@ -258,51 +297,132 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
+            for (SocketAddress address: addresses) {
+                ChannelFuture future = b.bind(address);
+                map.put(future, address);
+            }
+            return map;
+          }
+        }
+    );
+    Map<ChannelFuture, SocketAddress> channelFutures =
+        bindCallFuture.awaitUninterruptibly().getNow();
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:315]: This produces the wrong results if the future hasn't failed and just hasn't completed yet.

Comment (YifeiZhuang) [reply netty/src/main/java/io/grpc/netty/NettyServer.java:337]: Could you help elaborate on why awaitUninterruptibly() is not enough to have the result ready when getNow() is called? Thanks.

Comment (ejona86) [reply netty/src/main/java/io/grpc/netty/NettyServer.java:337]: I think I was probably blind to the `awaitUninterruptibly()`. This looks fine.

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 18]
```diff
@@ -78,7 +87,7 @@
   private EventLoopGroup bossGroup;
   private EventLoopGroup workerGroup;
   private ServerListener listener;
-  private Channel channel;
+  private volatile ChannelGroup channelGroup;
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:90]: Why is this volatile? It actually looks like it could be `final`

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 19]
```diff
@@ -258,51 +297,101 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
+                  public void operationComplete(ChannelFuture future) throws Exception {
+                    channelz.addListenSocket(listenSocketStats);
+                    listenSocketStatsList.add(listenSocketStats);
+                  }
+                });
+                ChannelFutureListener closeListener = new ChannelFutureListener() {
+                  @Override
+                  public void operationComplete(ChannelFuture future) throws Exception {
+                    if (!future.isSuccess()) {
+                      log.log(Level.WARNING, "Error closing server channel", future.cause());
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:323]: Drop this. This listener isn't responsible for closing, just interested to know when it happens. The code that actually closes should be handling this error.

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 20]
```diff
@@ -258,51 +297,101 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
+          channelGroup.close().awaitUninterruptibly();
+          throw new IOException(String.format("Failed to bind to address %s",
+              entry.getKey()), future.cause());
+        }
       }
-    });
+    }
+    for (Map<ChannelFuture, ChannelFutureListener> channelListener: channelFutures.values()) {
+      for (Map.Entry<ChannelFuture, ChannelFutureListener> entry: channelListener.entrySet()) {
+        entry.getKey().channel().closeFuture().removeListener(entry.getValue());
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:359]: Why do this? If you just leave the listener in-place, then you don't have to communicate it back to this thread (in the awkward map) and you don't need to call removeListenSocket within shutdown().

File: netty/src/main/java/io/grpc/netty/NettyServer.java (hunk 18 above)
Comment (YifeiZhuang) [reply netty/src/main/java/io/grpc/netty/NettyServer.java:90]: channel is added to channelGroup in the eventloop task when bind.

Comment (ejona86) [reply netty/src/main/java/io/grpc/netty/NettyServer.java:90]: Volatile has no effect on that. Volatile only controls the behavior of the variable's direct value. Since the value here never changes, it doesn't need to be volatile.

The issue you mention is impacted by whether ChannelGroup itself is thread-safe or not.

File: netty/src/main/java/io/grpc/netty/NettyServer.java (hunk 20 above)
Comment (YifeiZhuang) [reply netty/src/main/java/io/grpc/netty/NettyServer.java:359]: The close listener installed on the channel will stay with it in the entire lifetime, so if we do not remove it, then it conflicts with removeListenSocket at shutdown(), isn't it?

Comment (ejona86) [reply netty/src/main/java/io/grpc/netty/NettyServer.java:359]: I agree, but it seems simpler to remove the removal in shutdown().

File: netty/src/main/java/io/grpc/netty/NettyServer.java (hunk 18 above)
Comment (YifeiZhuang) [reply netty/src/main/java/io/grpc/netty/NettyServer.java:90]: It is supposed to be thread-safe and it is explicitly documented, likely we can make it final. (weird that i didn't get consistent result in my debug.)

Comment (ejona86) [reply netty/src/main/java/io/grpc/netty/NettyServer.java:90]: I'm a bit concerned that discussion of "final" seemed to be as a result of thread-safety of ChannelGroup. The two are completely separate; one has no impact on the other.

Comment (YifeiZhuang) [reply netty/src/main/java/io/grpc/netty/NettyServer.java]: Understood. Just as a result of that fact that there is no `final + volatile`.

File: netty/src/main/java/io/grpc/netty/NettyServer.java (hunk 20 above)
Comment (YifeiZhuang) [reply netty/src/main/java/io/grpc/netty/NettyServer.java]: I know where the conflict was from. Left the individual channel close listener removal there and the group close listener removal is ok to run again because of the emptiness check.

**Review by dapengzhang0 — APPROVED** [review]
LGTM

File: netty/src/main/java/io/grpc/netty/NettyServer.java (hunk 20 above)
Comment (ejona86) [reply netty/src/main/java/io/grpc/netty/NettyServer.java]: But you _don't need_ the group close listener removal! And setting `listenSocketStatsList = null` doesn't add much value and makes the code more complex. Why should this code even exist if the other code will handle the cleanup?

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 21]
```diff
@@ -97,10 +105,14 @@
   private final TransportTracer.Factory transportTracerFactory;
   private final InternalChannelz channelz;
   // Only modified in event loop but safe to read any time.
-  private volatile InternalInstrumented<SocketStats> listenSocketStats;
+  private volatile List<InternalInstrumented<SocketStats>> listenSocketStatsList =
```
[the remark this answers is not in the document]
Comment (YifeiZhuang) [reply netty/src/main/java/io/grpc/netty/NettyServer.java:108]: It can be modified in different channels' eventloop, and there is only a single one, so it looks fine at write. And the socketStats can be transient, so we return a snapshot on the direct read. I see the `getListenSocketStatsList` returned the reference directly which might cause problems as it is mutable, we can return an immutable list, or create a new copy then return.

[the remark this answers is not in the document]
Comment (YifeiZhuang) [reply netty/src/main/java/io/grpc/netty/NettyServer.java:108]: Actually each server channel's eventloop can be different.

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 22]
```diff
@@ -364,6 +368,7 @@ public void operationComplete(ChannelGroupFuture future) throws Exception {
               }
               sharedResourceReferenceCounter.release();
               protocolNegotiator.close();
+              listenSocketStatsList = null;
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:371]: Assign to `Collections.emptyList()` to avoid null pointer concerns?

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 23]
```diff
@@ -348,6 +349,9 @@ public void operationComplete(ChannelFuture future) throws Exception {
             entry.getKey()), future.cause());
       }
     }
+    synchronized (NettyServer.this) {
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:352]: This synchronization does nothing.

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 24]
```diff
@@ -309,20 +314,16 @@ public void operationComplete(ChannelFuture future) throws Exception {
                   public void operationComplete(ChannelFuture future) throws Exception {
                     channelz.addListenSocket(listenSocketStats);
                     synchronized (NettyServer.this) {
-                      listenSocketStatsList.add(listenSocketStats);
+                      socketStats.add(listenSocketStats);
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:317]: This is not thread-safe, as it can happen after `future.awaitUninterruptibly().isSuccess()` returns.

We had discussed needing to have another future that is completed right after this synchronized block, and pass that future instead in `bindFutures`. But I actually think it'd be easier to move this code (and listenSocketStats creation) to the `channelFutures.entrySet()` for loop directly. At that location, the entire contents of this `ChannelFutureListener` could be called directly and the synchronized block would be unnecessary. It'd be best to move `future.channel().closeFuture().addListener()` for `channelz.removeListenSocket()` to the for loop as well so they stay paired (but it would stay as a listener).

Comment (YifeiZhuang) [reply netty/src/main/java/io/grpc/netty/NettyServer.java:317]: Why synchronization block (L352) is not enough to ensure all write to the arraylist is sequentially done and the read is also up to date?

Comment (ejona86) [reply netty/src/main/java/io/grpc/netty/NettyServer.java:317]: Because this listener may be executed after that block.

Comment (YifeiZhuang) [reply netty/src/main/java/io/grpc/netty/NettyServer.java:317]: k, could you remind me why we originally have [this](https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/NettyServer.java#L271-L276) in eventloop?

Comment (ejona86) [reply netty/src/main/java/io/grpc/netty/NettyServer.java:317]: I don't think it is necessary. What is necessary is that it is executed after the bind completes. It appears it could have been simplified when https://github.com/grpc/grpc-java/commit/49c8bdb60ac8783eda098aa42375b837342edde3 was done, but wasn't noticed at the time (nor in https://github.com/grpc/grpc-java/commit/b8474d61c966f60cc0bad4e5abaca351e90afc9d)

Comment (YifeiZhuang) [reply netty/src/main/java/io/grpc/netty/NettyServer.java]: Thanks. I put it inline, but I am actually not hundred percent sure if `L341` is publish-safe. Reference assignment is fine, and because the list is modified in a single thread, it looks other thread would `get()` what this thread can see by the time unmodifiable view is created, which might be safe.

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 25]
```diff
@@ -258,51 +293,77 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
+      throw new IOException(String.format("Failed to bind to addresses %s",
+          addresses), bindCallFuture.cause());
     }
-    channel = future.channel();
-    channel.eventLoop().execute(new Runnable() {
-      @Override
-      public void run() {
-        listenSocketStats = new ListenSocket(channel);
-        channelz.addListenSocket(listenSocketStats);
+    final List<InternalInstrumented<SocketStats>> socketStats = new ArrayList<>();
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java:319]: nit: no need for final.

**Review by ejona86 — APPROVED** [review]

File: netty/src/main/java/io/grpc/netty/NettyServer.java (hunk 24 above)
Comment (ejona86) [reply netty/src/main/java/io/grpc/netty/NettyServer.java]: It looks fine. Yes, the volatile write publishes any writes that occurred before it on the same thread.
